Skip the device if the battery is missing

stuarthayhurst opened the pull request:
Some devices (such as the wireless pen on the Lenovo Yoga 7 Gen 7) report a battery even if it's missing.
This checks both state and icon name, as some devices may not report the battery state, while still having a battery present. If you want, I can add that line in as a comment.

Let me know if you want any changes :)

stuarthayhurst commented:
I forgot to hide devices that correctly report when a battery is missing, fixed that
I also now check UPower.DeviceState.UNKNOWN instead of 0, and added the logic as a comment

stuarthayhurst commented:
This is causing some issues for my Logitech G502 mouse, I'll submit a fix for that ASAP

stuarthayhurst commented:
Sorted all the issues I could find, working on all my devices now. When devices are created, they're given an icon regardless of if there's a battery or not. After, if there's no battery attached or one is added / removed, the device releases a signal. The WirelessHID class picks this signal up, and then removes / adds the icon.

I'll happily clear up any confusion, as I'm not sure I gave a very good explanation.

vchlum commented on wirelesshid.js:
```diff
@@ -120,12 +124,31 @@ var HID = GObject.registerClass({
+
+        //Some devices report 'present' as true, even if no battery is present
+        //Instead, check the state and icon name to try and find these
+        if (this.device.state === UPower.DeviceState.UNKNOWN && this.device.iconName === "battery-missing-symbolic") {
```
My Logitech Mouse MX and Logitech MX Keys both have this Unknown state and also the missing icon...

Like this:
```
  keyboard
    present:             yes
    rechargeable:        no
    state:               unknown
    warning-level:       none
    percentage:          100%
    icon-name:          'battery-missing-symbolic'

  mouse
    present:             yes
    rechargeable:        no
    state:               unknown
    warning-level:       none
    percentage:          65%
    icon-name:          'battery-missing-symbolic'


```
Those  two devices are the reason why I created this extension:-) So this condition is definitely not wanted. Obviously, we can not rely on the attributes. I am afraid removing the condition will break your feature, right?

stuarthayhurst replied:
It would break the detection of receivers with removable devices, yes.

stuarthayhurst replied:
Should be sorted by the latest commit

vchlum reviewed (changes requested):
Thank you so much for your contribution. Unfortunately, the unknown attributes of devices mean 'unknown'. Not 'disabled' nor 'unavailable'. We can not use it to disable devices. I am not sure if this feature can be achieved reliably without user input.

stuarthayhurst commented:
That's frustrating. Is that when they're both turned on as well? If that's the case, I'm happy to make add a settings window for all supported GNOME versions, and make this a user toggleable feature.

vchlum commented:
Yes, both the mouse and keyboard are in use while this state is shown. If you are willing to create settings it would be much appreciated! I intend to do it by myself but due to other projects, it is not on top of the queue. So any help is great!:-)

stuarthayhurst commented:
No worries, I've got 3 of my own extensions so I'm quite familiar with the settings windows for GNOME 3.36 -> GNOME 43. I'll get this sorted and add a small section to the README on what the setting does :)

stuarthayhurst commented:
I rewrote the code to handle the menu entries and icons, should fix a couple bugs I noticed. Most importantly, each device is completely responsible for its own icon now.

I also added the settings window and setting. New dependencies (`libgtk4-bin` and `libglib2.0-bin`) are needed to convert GTK3 files to GTK4, and compile the gsettings schemas. The code for this (`prefs.js`) is in my own programming style, I can try adjust it to match yours if you'd like.

The new script (`[update-ui.sh](http://update-ui.sh/)`) only needs to be run whenever the `ui/prefs.ui` file is modified. `ui/prefs-gtk4.ui` shouldn't be edited by hand or by a UI editor, it should only be generated. It's technically a build file, but committing it prevents another dependency being added to build it. It only changes when `ui/prefs.ui` changes anyway.

However, there's still a bit left to do:
 - Refresh the devices when settings change
 - Test the settings window on GNOME 3.36 -> GNOME 41
 - Document the new dependencies
 - Document the new setting

There are also potentially 2 bugs left:
 - Icon disappears sometimes after unlocking the device
 - An empty indicator is present when no devices are left
These issues don't always appear, so I *think* I've fixed them, but I've not been running the fix for long enough to be confident yet.

I also have a few code improvements to my new code left, as long as you're happy with the current logic.

stuarthayhurst commented:
I based the code for the settings window on one of my other extensions, [here](https://github.com/stuarthayhurst/alphabetical-grid-extension/blob/master/extension/prefs.js). If you add more settings in future, you can probably use that extension as a reference, or just @ me if my code wasn't very clear.

vchlum commented:
* Refresh the devices when settings change -> DONE
* Icon disappears sometimes after unlocking the device -> I locked and unlocked my notebook several times and the icons are always there, I suppose it is not an issue.
I will do more testing and release it soon.

stuarthayhurst commented:
Ah thanks, saves me adding that :)

I'm pretty sure I've sorted all the bugs in this PR, it's just the disappearing icon that was a bit hard to reproduce, but that should be sorted. If it comes back I'll file an issue and try find a fix.

stuarthayhurst commented:
@vchlum 
I noticed all ELAN devices are blocked, but the reason (#6) should be covered by this PR. Perhaps it's worth reverting a620afceef43040ea9fe5dd2a3d72def1e0d8b56?

Also, if you need help testing the extension on different GNOME versions, LMK which and I'll create some VMs for testing.

vchlum commented:
Hi @stuarthayhurst, first of all, Marry Christmas!

Sorry for not getting back to you sooner, I tested the gnome 3.36. It looks good, so the new version should soon be available.

It is not that easy... ELAN on my device (ASUS Zenbook)  has no battery available in upower but upower shows it. I do not want to show ELAN but I want to show other devices with unknown batteries. IMO, the best solution would be to allow disabling particular devices.

stuarthayhurst commented:
Happy Christmas! No worries about the delay, I figured people would be busy around the holidays anyway :)

Good to hear it's working on GNOME 3.36, and I suppose the ELAN devices are similar to the Wacom device on my Lenovo Yoga 7 Gen 7. I suppose listing the devices in the settings menu, then allowing the users to blacklist specific devices could work.

vchlum commented:
Yes, exactly it is the stylus battery. Mine is actually this [one](https://github.com/linuxwacom/wacom-hid-descriptors/tree/master/Asus%20Zenbook%20Flip%20S13%20UX371). We could also add a possibility to enable/disable ELAN devices with a separate switch. It is a quick solution. The proper one could come later. If ever need... I think there is no harm in it. I have already started.
